File: PVFC.py
import random
import time

def char_to_num(a):
    if not isinstance(a, str):
        raise ValueError("Input must be a string.")
    return ''.join(str(ord(b)).zfill(3) for b in a)

def num_to_char(a):
    if not isinstance(a, str) or len(a) % 3 != 0:
        raise ValueError("Input must be a string with a length that is a multiple of 3.")
    return ''.join(chr(int(a[b:b+3])) for b in range(0, len(a), 3))

def generate_key(length):
    random.seed(int(time.time() * (random.randint(1000, 64000) * random.randint(1, 255))))
    random_bytes = [random.randint(0, 255) for _ in range(length * 16)]
    key = custom_hash(random_bytes) % (2 ** (length * 8))
    key = apply_bitwise_operations(key)
    key_hex = hex(key)[2:].zfill(length * 2)
    return key_hex

def custom_hash(random_bytes):
    hash_value = 0
    for byte in random_bytes:
        hash_value ^= byte
        hash_value = (hash_value << 5) - hash_value
    return hash_value

def apply_bitwise_operations(key):
    key = key ^ (key << 13) & 0xFFFFFFFFFFFFFFFF
    key = key ^ (key >> 7) & 0xFFFFFFFFFFFFFFFF
    key = key ^ (key << 3) & 0xFFFFFFFFFFFFFFFF
    return key

def generate_keys(num_keys, key_length):
    return [generate_key(key_length) for _ in range(num_keys)]

c = "0123456789abcdefghijklmnopqrstuvwxyz"

def encrypt(x, y):
    if not isinstance(x, str) or not isinstance(y, list) or not all(isinstance(k, str) for k in y):
        raise ValueError("Invalid input for encryption.")
    z = ''
    a = max(len(b) for b in y)
    for d in range(0, a, 2):
        for e in y:
            if d < len(e):
                z += e[d:d+2]
    
    f = [int(z[g:g+2], 16) for g in range(0, len(z), 2)]
    h = [int(x[i:i+3]) for i in range(0, len(x), 3)]
    
    for j in range(len(h)):
        k = h[j]
        k = (k + f[j % len(f)]) % 256
        k = k ^ f[(j + 1) % len(f)]
        l = (f[j % len(f)] + j) % 8
        k = ((k << l) | (k >> (8 - l))) & 0xFF
        k = k ^ ((f[j % len(f)] + j) % 256)
        h[j] = k
    
    m = ''.join(f"{n:03d}" for n in h)
    o = list(m)
    
    for p in range(len(m)):
        q = (p + f[p % len(f)]) % len(m)
        o[p], o[q] = o[q], o[p]
    
    r = ''.join(o)
    s = ""
    
    for t in range(0, len(r), 3):
        u = r[t:t+3]
        v = int(u)
        s += c[v % 36]
        s += c[(v // 36) % 36]
    
    return s

def decrypt(x, y):
    if not isinstance(x, str) or not isinstance(y, list) or not all(isinstance(k, str) for k in y):
        raise ValueError("Invalid input for decryption.")
    z = ""
    
    for a in range(0, len(x), 2):
        b = x[a:a+2]
        d = c.index(b[0]) + c.index(b[1]) * 36
        z += f"{d:03d}"
    
    e = ''
    f = max(len(g) for g in y)
    for h in range(0, f, 2):
        for i in y:
            if h < len(i):
                e += i[h:h+2]
    
    j = [int(e[k:k+2], 16) for k in range(0, len(e), 2)]
    l = list(z)
    
    for m in range(len(z) - 1, -1, -1):
        n = (m + j[m % len(j)]) % len(z)
        l[m], l[n] = l[n], l[m]
    
    o = ''.join(l)
    p = [int(o[q:q+3]) for q in range(0, len(o), 3)]
    
    for r in range(len(p) - 1, -1, -1):
        s = p[r]
        s = s ^ ((j[r % len(j)] + r) % 256)
        t = (j[r % len(j)] + r) % 8
        s = ((s >> t) | (s << (8 - t))) & 0xFF
        s = s ^ j[(r + 1) % len(j)]
        s = (s - j[r % len(j)]) % 256
        p[r] = s
    
    u = ''.join(f"{v:03d}" for v in p)
    return u

if __name__ == "__main__":
    input_str = input("Enter a string: ")
    output_num = char_to_num(input_str)
    print("Numerical Representation:", output_num)

    num_keys = 4
    key_length = 32
    keys = generate_keys(num_keys, key_length)
    print("Generated Keys:", keys)

    encrypted_hex = encrypt(output_num, keys)
    print("Encrypted:", encrypted_hex)

    decrypted_num = decrypt(encrypted_hex, keys)
    print("Decrypted Numerical Representation:", decrypted_num)

    decrypted_str = num_to_char(decrypted_num)
    print("Decrypted String:", decrypted_str)

    if input_str == decrypted_str:
        print("Success: The original string matches the decrypted string.")
    else:
        print("Error: The original string does not match the decrypted string.")

File: PVFC.js
function charToNum(a) {
    return Array.from(a).map(b => b.charCodeAt(0).toString().padStart(3, '0')).join('');
}

function numToChar(a) {
    return Array.from({ length: a.length / 3 }, (_, i) => String.fromCharCode(parseInt(a.slice(i * 3, i * 3 + 3)))).join('');
}

function generateKeys(a, b) {
    return Array.from({ length: a }, () => generateKey(b));
}

function generateKey(a) {
    return Array.from({ length: a }, () => Math.floor(Math.random() * 256).toString(16).padStart(2, '0')).join('');
}

const c = "0123456789abcdefghijklmnopqrstuvwxyz";

function encrypt(x, y) {
    let z = '';
    const a = Math.max(...y.map(b => b.length));
    for (let d = 0; d < a; d += 2) {
        for (const e of y) {
            if (d < e.length) {
                z += e.slice(d, d + 2);
            }
        }
    }

    const f = Array.from({ length: z.length / 2 }, (_, g) => parseInt(z.slice(g * 2, g * 2 + 2), 16));
    const h = Array.from({ length: x.length / 3 }, (_, i) => parseInt(x.slice(i * 3, i * 3 + 3)));

    for (let j = 0; j < h.length; j++) {
        let k = h[j];
        k = (k + f[j % f.length]) % 256;
        k = k ^ f[(j + 1) % f.length];
        const l = (f[j % f.length] + j) % 8;
        k = ((k << l) | (k >>> (8 - l))) & 0xFF;
        k = k ^ ((f[j % f.length] + j) % 256);
        h[j] = k;
    }

    const m = h.map(n => n.toString().padStart(3, '0')).join('');
    const o = m.split('');

    for (let p = 0; p < m.length; p++) {
        const q = (p + f[p % f.length]) % m.length;
        [o[p], o[q]] = [o[q], o[p]];
    }

    const r = o.join('');
    let s = "";

    for (let t = 0; t < r.length; t += 3) {
        const u = r.slice(t, t + 3);
        const v = parseInt(u);
        s += c[v % 36];
        s += c[Math.floor(v / 36) % 36];
    }

    return s;
}

function decrypt(x, y) {
    let z = "";

    for (let a = 0; a < x.length; a += 2) {
        const b = x.slice(a, a + 2);
        const d = c.indexOf(b[0]) + c.indexOf(b[1]) * 36;
        z += d.toString().padStart(3, '0');
    }

    let e = '';
    const f = Math.max(...y.map(g => g.length));
    for (let h = 0; h < f; h += 2) {
        for (const i of y) {
            if (h < i.length) {
                e += i.slice(h, h + 2);
            }
        }
    }

    const j = Array.from({ length: e.length / 2 }, (_, k) => parseInt(e.slice(k * 2, k * 2 + 2), 16));
    const l = z.split('');

    for (let m = l.length - 1; m >= 0; m--) {
        const n = (m + j[m % j.length]) % l.length;
        [l[m], l[n]] = [l[n], l[m]];
    }

    const o = l.join('');
    const p = Array.from({ length: o.length / 3 }, (_, q) => parseInt(o.slice(q * 3, q * 3 + 3)));

    for (let r = p.length - 1; r >= 0; r--) {
        let s = p[r];
        s = s ^ ((j[r % j.length] + r) % 256);
        const t = (j[r % j.length] + r) % 8;
        s = ((s >>> t) | (s << (8 - t))) & 0xFF;
        s = s ^ j[(r + 1) % j.length];
        s = (s - j[r % j.length] + 256) % 256;
        p[r] = s;
    }

    return p.map(v => v.toString().padStart(3, '0')).join('');
}

const plaintext = "Hello, World!";
const numKeys = 3;
const keyLength = 16;

console.log("Original text:", plaintext);

const numericText = charToNum(plaintext);
console.log("Numeric representation:", numericText);

const keys = generateKeys(numKeys, keyLength);
console.log("Generated keys:", keys);

const encryptedText = encrypt(numericText, keys);
console.log("Encrypted text:", encryptedText);

const decryptedNumeric = decrypt(encryptedText, keys);
const decryptedText = numToChar(decryptedNumeric);
console.log("Decrypted text:", decryptedText);

console.log("Decryption successful:", plaintext === decryptedText);
